### ISSUE TITLE
refactor into several npm modules

### ISSUE THREAD

**Issue by tj** [componentjs/builder.js#78]
to help people build their own builders if they dont want to use this one
- dep tree builder with lookup paths
- module definition with sourceurl option
- require related helpers
- url rewriting

**Comment by thehydroimpulse** [comment]
@visionmedia Also, it would be awesome if you provided hooks to implement custom module loaders. Right now (unless i'm missing something) i'm required to hack around the current builder and override a bunch of methods.

**Comment by jonathanong** [comment]
are you working on a tree builder yet? i was thinking of writing one that fixes https://github.com/component/builder.js/issues/109

**Comment by tj** [comment]
yeah we'll need to separate the tree building to do a first pass and map all the requires so we can replace them without the aliases
